backends-velox/src/main/scala/org/apache/gluten/execution/ColumnarCollectLimitExec.scala

```diff
@@ -0,0 +1,153 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.execution
+
+import org.apache.gluten.backendsapi.BackendsApiManager
+import org.apache.gluten.columnarbatch.{ColumnarBatches, VeloxColumnarBatches}
+import org.apache.gluten.extension.columnar.transition.Convention
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.serializer.Serializer
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.SinglePartition
+import org.apache.spark.sql.execution.{ShuffledColumnarBatchRDD, SparkPlan}
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLShuffleWriteMetricsReporter}
+import org.apache.spark.sql.metric.SQLColumnarShuffleReadMetricsReporter
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+case class ColumnarCollectLimitExec(
+    limit: Int,
+    child: SparkPlan
+) extends ColumnarCollectLimitBaseExec(limit, child) {
+
+  override def batchType(): Convention.BatchType =
+    BackendsApiManager.getSettings.primaryBatchType
+
+  /**
+   * Returns an iterator that yields up to `limit` rows in total from the input partitionIter.
+   * Either retain the entire batch if it fits within the remaining limit, or prune it if it
+   * partially exceeds the remaining limit.
+   */
+  private def collectLimitedRows(
+      partitionIter: Iterator[ColumnarBatch],
+      limit: Int
+  ): Iterator[ColumnarBatch] = {
+    if (partitionIter.isEmpty) {
+      return Iterator.empty
+    }
+    new Iterator[ColumnarBatch] {
+
+      private var rowsCollected = 0
+      private var nextBatch: Option[ColumnarBatch] = None
+
+      override def hasNext: Boolean = {
+        nextBatch.isDefined || fetchNext()
+      }
+
+      override def next(): ColumnarBatch = {
+        if (!hasNext) {
+          throw new NoSuchElementException("No more batches available.")
+        }
+        val batch = nextBatch.get
+        nextBatch = None
+        batch
+      }
+
+      /**
+       * Attempt to fetch the next batch from the underlying iterator if we haven't yet hit the
+       * limit. Returns true if we found a new batch, false otherwise.
+       */
+      private def fetchNext(): Boolean = {
+        if (rowsCollected >= limit || !partitionIter.hasNext) {
+          return false
+        }
+
+        val currentBatch = partitionIter.next()
+        val currentBatchRowCount = currentBatch.numRows()
+        val remaining = limit - rowsCollected
+
+        ColumnarBatches.retain(currentBatch)
```

Comment:
When we retain currentBatch, the last operator will not close it, but the next operator only closes the sliced batch, the original parent batch is not closed in time, I suppose it is closed after the query ends up. Do we need to add a `SliceBatch` like `ComposeBatch` to maintain the same lifecycle of batches? Then the next operator closes both of them. @zhztheplayer

Reply:
The sliced batch is a view of input batch without no copy.

Reply:
> The sliced batch is a view of input batch without no copy.

Doesn't Velox handle the referencing internally? I assume it's easy in C++ with shared pointers.

https://github.com/facebookincubator/velox/blob/48d21bfd74be313d54b6cd33530525575ac44717/velox/vector/FlatVector-inl.h#L417-L438

What happens if we don't retain the batch?

Reply:
It returns a BufferView, https://github.com/facebookincubator/velox/blob/48d21bfd74be313d54b6cd33530525575ac44717/velox/buffer/Buffer.h#L181
So the original Buffer value should be keepped.

Reply:
Thanks for your investigation, the sliced BufferView will hold the shared pointer of value Buffer, so we don't need to retain the original Batch.

Reply:
Could you please update the `retain` and rebase? Then we can merge it. @ArnavBalyan